### PR TITLE
Fixed Code execution in swig

### PR DESCRIPTION
**Pull request by alromh87**
### 📊 Metadata *

SWIG is a compiler that integrates C and C++ with languages including Perl, Python, Tcl, Ruby, PHP, Java, C#, D, Go, Lua, Octave, R, Scheme (Guile, MzScheme/Racket), Scilab, Ocaml. SWIG can also export its parse tree into XML.

### ⚙️ Description *

One of the python tools of swig include a mkdist.py script. This script takes in arguments and execute it without sanitation leading to Arbitrary code execution. Arguments are now sanitized before execution

### 💻 Technical Description *

Shell escaping characters are filtered using regexp

### 🐛 Proof of Concept (PoC) *

1. Download swig
2. Execute POC
```sh
python Tools/mkdist.py "a; touch hacked #"
```
3. Hacked file wil be created in working directory after code execution

Captura de pantalla de 2020-09-10 15-04-12

### 🔥 Proof of Fix (PoF) *

After fix no file is created

Captura de pantalla de 2020-09-10 15-03-44

### 👍 User Acceptance Testing (UAT)

Comands can be executed normally

### CODE REVIEW

**Review by Mik317 — CHANGES_REQUESTED** [review]
The `fix` is bypassable since `|` isn't checked 😢 

PoC: 
image

Cheers,
Mik

**Comment by alromh87** [comment]
Thanks Mik I updated for best performance  in python

**Review by Mik317 — CHANGES_REQUESTED** [review]
Still bypassable using `backticks` :

image

Cheers,
Mik

**Comment by alromh87** [comment]
Thanks

File: Tools/mkdist.py [hunk 1]
```diff
@@ -3,11 +3,15 @@
...
 import os
 import subprocess
+import re
 
 def failed():
     print("mkdist.py failed to complete")
     sys.exit(2)
 
+def clean_arg(arg):
+    return arg.replace('`','').replace('\'','').replace('"','').replace(';','').replace('&','').replace(',','').replace('\\','').replace('|','')
```
Comment (Mik317) [inline Tools/mkdist.py]: Clean also `${}` and `$()` since could result in `RCE`

**Review by Mik317 — CHANGES_REQUESTED** [review]
Hi @alromh87 :smile:,
thanks for the `minor fix` :+1: 
I identified another small `bypass` which makes possible `RCE` (more info on how resolve in the comment):

Screenshot from 2020-09-10 23-39-35

Cheers,
Mik

File: Tools/mkdist.py (hunk 1 above)
Comment (alromh87) [reply Tools/mkdist.py]: This is a tricky one, thanks for your help debbuging

File: Tools/mkdist.py [hunk 2]
```diff
@@ -10,7 +10,7 @@ def failed():
     sys.exit(2)
 
 def clean_arg(arg):
-    return arg.replace('`','').replace('\'','').replace('"','').replace(';','').replace('&','').replace(',','').replace('\\','').replace('|','')
+    return arg.replace('`','').replace('\'','').replace('"','').replace(';','').replace('&','').replace(',','').replace('\\','').replace('|','').replace('$','')
```
Comment (Mik317) [inline Tools/mkdist.py:13]: Still bypassable :cry: 
The `(command)` format is possible in `Windows`, leading to `RCE` again

**Review by Mik317 — CHANGES_REQUESTED** [review]
The fix is still bypassable using another syntax.
I added more info in the comment :smile:

Cheers,
Mik

**Comment by alromh87** [comment]
Implemented a cleaner fix

https://github.com/418sec/swig/pull/2

**Pull request closed without merging**